substrate/frame/staking/src/pallet/mod.rs

```diff
@@ -1980,6 +1983,115 @@ pub mod pallet {
 
 			Ok(Some(T::WeightInfo::deprecate_controller_batch(controllers.len() as u32)).into())
 		}
+
+		/// Restores the state of a ledger which is in an inconsistent state.
+		///
+		/// The requirements to restore a ledger are the following:
+		/// * The stash is bonded;
+		/// * If the stash has an associated ledger, its state must be inconsistent;
+		/// * If the ledger is not corrupted *but* its staking lock is out of sync.
+		///
+		/// The `maybe_*` input parameters will overwrite the corresponding data and metadata of the
+		/// ledger associated with the stash. If the input parameters are not set, the ledger will
+		/// be reset with the current values.
+		///
+		/// Upon successful execution, this extrinsic will *recreate* the ledger based on its past
+		/// state and/or the `maybe_controller` and `maybe_total` input parameters.
+		#[pallet::call_index(29)]
+		#[pallet::weight(T::WeightInfo::restore_ledger())]
+		pub fn restore_ledger(
+			origin: OriginFor<T>,
+			stash: T::AccountId,
+			maybe_controller: Option<T::AccountId>,
+			maybe_total: Option<BalanceOf<T>>,
+		) -> DispatchResultWithPostInfo {
+			T::AdminOrigin::ensure_origin(origin)?;
+
+			// closure to call before return from extrinsic to ensure the restore was successful.
+			let integrity_checks = || {
+				let lock = T::Currency::balance_locked(crate::STAKING_ID, &stash);
+
+				if let Some(ledger) = Bonded::<T>::get(&stash).and_then(Ledger::<T>::get) {
+					ensure!(lock == ledger.total, Error::<T>::BadState);
+					ensure!(Payee::<T>::get(&stash).is_some(), Error::<T>::BadState);
+					ensure!(ledger.stash == stash, Error::<T>::BadState);
+				} else {
+					ensure!(lock == Zero::zero(), Error::<T>::BadState);
+				}
+
+				Ok::<_, Error<T>>(())
+			};
+
+			let current_lock = T::Currency::balance_locked(crate::STAKING_ID, &stash);
+
+			let (new_controller, new_total) = match Self::inspect_bond_state(&stash) {
+				Ok(LedgerIntegrityState::Corrupted) => {
+					let new_controller = maybe_controller.unwrap_or(stash.clone());
+
+					let new_total = if let Some(new_total) = maybe_total {
+						// enforce lock == ledger.amount.
+						T::Currency::set_lock(
+							crate::STAKING_ID,
+							&stash,
+							new_total,
+							WithdrawReasons::all(),
+						);
+						new_total
+					} else {
+						current_lock
+					};
+
+					Ok((new_controller, new_total))
+				},
+				Ok(LedgerIntegrityState::CorruptedKilled) => {
+					if current_lock == Zero::zero() {
```

Comment:
Is there any case like this currently? If lock is zero and ledger is 0 why do we need to restore? Seems good state?

Affected user can just do a new bond. Otherwise this will allow staking admin to bond any non bonded account.

Reply:
>  If lock is zero and ledger is 0 why do we need to restore? Seems good state?

Not really. In this case, there is a lingering bonded entry for the stash. So basically:

```
Bonded(stash) = C;
Lock(stash) = 0;
Ledger(C) = None;
```

This may happen in the case when a corrupted ledger has been killed, e.g. [like in this case]https://hackmd.io/DLb5jEYWSmmvqXC9ae4yRg#/6).

Reply:
In [this case](https://hackmd.io/DLb5jEYWSmmvqXC9ae4yRg#/6) `Lock(stash)` is equal to previous `ledger.total`.

Reply:
@Ank4n here I think it's more clear why we should not clear the bond in this case bur rather recover it https://github.com/paritytech/polkadot-sdk/pull/3866/files#diff-24281cfd6db327807105722f46ef26850905bc77759978a2fff784cb9379817cR7578